### PR TITLE
[Chains] Make context optional, various improvements.

### PR DESCRIPTION
**Pull request by marius-baseten**
* Make context argument optional and, if used, trailing.
* Rename `run` -> `run_remote`, `provide` -> `depends`.
* Configurable predict concurrency.
* Better detection of invalid chainelt instatiations.
* Add more unittests - covering chainlet instatiations warnings.

### CODE REVIEW

File: truss-chains/tests/chains_test.py [hunk 1]
```diff
@@ -46,16 +48,110 @@ async def test_chain_local():
...
+            definitions.ChainsRuntimeError,
+            match="Chainlets cannot be naively instantiated",
+        ):
+            await entrypoint().run_remote(length=20, num_partitions=5)
+
+
+# Chainlet Initialization Guarding #####################################################
+
+
+def test_raises_without_depends():
```
Comment (squidarth) [inline truss-chains/tests/chains_test.py:71]: These tests are awesome

File: truss-chains/truss_chains/definitions.py [hunk 2]
```diff
@@ -249,30 +277,24 @@ def get_baseten_api_key(self) -> str:
...
 
-    user_config: UserConfigT = pydantic.Field(default=None)
+    user_config: UserConfigT
     chainlet_to_service: Mapping[str, ServiceDescriptor] = {}
 
 
-class ABCChainlet(Generic[UserConfigT], abc.ABC):
-    remote_config: ClassVar[RemoteConfig]
+class ABCChainlet(abc.ABC):
+    remote_config: ClassVar[RemoteConfig] = RemoteConfig(docker_image=DockerImage())
```
Comment (squidarth) [inline truss-chains/truss_chains/definitions.py:285]: just a note: I think even if we do set a default here, our examples should have this filled out, so it's quicker to know how to modify

Comment (squidarth) [reply truss-chains/truss_chains/definitions.py:285]: One other default we could consider is having RemoteConfig() defaulting the `docker_image`

File: truss-chains/truss_chains/framework.py [hunk 3]
```diff
@@ -42,12 +43,22 @@ class _BaseProvisionMarker:
...
 
-class ContextProvisionMarker(_BaseProvisionMarker):
+class ContextDependencyMarker(_BaseProvisionMarker):
     def __str__(self) -> str:
         return f"{self.__class__.__name__}"
 
+    def __getattr__(self, item: str) -> Any:
+        logging.error(f"Attempting to access attribute `{item}` on `{self}`.")
+        raise definitions.ChainsRuntimeError(
+            "It seems `chains.depends_context()` was used, but not as an argument "
```
Comment (squidarth) [inline truss-chains/truss_chains/framework.py:53]: nice error

**Review by squidarth — APPROVED** [review]
Approving to unblock mostly, I will take a closer look tomorrow!

File: truss-chains/truss_chains/definitions.py (hunk 2 above)
Comment (marius-baseten) [reply truss-chains/truss_chains/definitions.py:285]: Agree, let's keep it in examples as a pointer.

The original idea for docker image as mandatory was to "force" users to be intentional, but not that the whole thing is optional anyway, we can also drop that requirement...

**Pull request merged**